src/id.rs

```diff
@@ -12,28 +14,105 @@ pub(crate) struct Id {
 }
 
 pub(crate) struct Allocator {
-    new_id: u32,
-    free_ids: Vec<Id>,
+    new_id: atomic::AtomicU32,
+    free_ids: spin::RwLock<IdCache>,
 }
 
 impl Allocator {
     pub fn new() -> Self {
-        Allocator { new_id: 0, free_ids: Vec::new() }
-    }
-
-    pub fn allocate(&mut self) -> Id {
-        // TODO: better balancing
-        match self.free_ids.pop() {
-            Some(id) => Id { id: id.id, generation: id.generation + 1 },
-            None => {
-                let id = self.new_id;
-                self.new_id += 1;
-                Id { id, generation: 0 }
-            }
+        Allocator { new_id: atomic::AtomicU32::new(0), free_ids: spin::RwLock::new(IdCache::default()) }
+    }
+
+    pub fn allocate(&self) -> Id {
+        self.free_ids.read().pop_atomic().map(|Id { id, generation }| Id { id, generation: generation + 1 }).unwrap_or_else(
+            || Id { id: atomic_increment(&self.new_id).expect("No entity left to allocate"), generation: 0 },
```

Comment:
Yes, this panics, and it would seem like the crate has a policy of not panicking.  However: the old code would also have panicked here due to integer overflow; and if you have `u32::MAX` Ids alive at once, you probably have bigger problems anyway (e.g. out of memory)